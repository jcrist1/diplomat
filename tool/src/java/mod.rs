use std::collections::HashSet;
use std::fs::File;
use std::io::Write;
use std::path::Path;

use diplomat_core::hir::TypeContext;

use crate::c2;
use crate::common::FileMap;

const TMP_C_DIR: &str = "tmp";
const LIBRARY: &str = "somelib"; // todo: build from conf. Ensure that name is not the same as any
                                 // type
const GROUP: &str = "dev.diplomattest"; // todo: config
const TMP_LIB_NAME: &str = "dev/diplomattest/somelib"; // todo: build from conf
const JAVA_DIR: &str = "src/main/java/";

pub fn run(
    tcx: &TypeContext,
    _conf_path: Option<&Path>,
    out_folder: &Path,
) -> std::io::Result<FileMap> {
    let files = FileMap::default();
    let mut context = c2::CContext::new(tcx, files, false);
    context.run();

    let errors = context.errors.take_all();

    if !errors.is_empty() {
        eprintln!("Found errors when generating c  code");
        for error in errors {
            eprintln!("\t{}: {}", error.0, error.1);
        }
    }

    let out_files = context.files.take_files();

    let tmp_path = out_folder.join(TMP_C_DIR);
    std::fs::create_dir(&tmp_path)?;
    let mut include_files = HashSet::new();
    for (subpath, text) in out_files {
        let out_path = tmp_path.join(&subpath);
        if !subpath.ends_with(".d.h") && subpath.ends_with(".h") {
            include_files.insert(subpath);
        }
        let parent = out_path
            .parent()
            .expect("Cannot create files at top level dir /");
        std::fs::create_dir_all(parent)?;
        let mut out_file = File::create(&out_path)?;
        out_file.write_all(text.as_bytes())?;
    }

    let lib_path = tmp_path.join(format!("{LIBRARY}.h"));

    let mut lib_file = File::create(&lib_path)?;
    for include in include_files {
        writeln!(lib_file, "#include \"{include}\"")?;
    }

    // jextract \
    //   --include-dir /path/to/mylib/include \
    //   --output src \
    //   --target-package org.jextract.mylib \
    //   --library mylib \
    //   /path/to/mylib/include/mylib.h

    let package = format!("{GROUP}.{LIBRARY}.ntv");
    let mut command = std::process::Command::new("jextract");
    command
        .arg("--include-dir")
        .arg(&tmp_path)
        .arg("--output")
        .arg(out_folder)
        .arg("--target-package")
        .arg(package)
        .arg("--library")
        .arg(LIBRARY)
        .arg(lib_path);

    println!("Running: {:?}", command);

    // todo: delete directory

    match command.output() {
        Err(err) => match err.kind() {
            std::io::ErrorKind::NotFound => {
                eprintln!("Check that jextract is in your path and all directories exist. See https://github.com/openjdk/jextract/blob/5715737be0a1a9de24cce3ee7190881cfc8b1350/doc/GUIDE.md");
                return Err(err);
            }
            _ => return Err(err),
        },
        Ok(ok) => {
            let stdout = String::from_utf8_lossy(&ok.stdout);
            println!("Output from jextract:\n{stdout}");

            let stderr = String::from_utf8_lossy(&ok.stderr);
            println!("Std Err from jextract:\n{stderr}");
        }
    }

    let files = FileMap::default();
    Ok(files)
}
