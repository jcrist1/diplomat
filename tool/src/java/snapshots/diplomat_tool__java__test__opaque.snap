---
source: tool/src/java/mod.rs
expression: res
---
package dev.diplomattest.somelib;

import dev.diplomattest.somelib.ntv.somelib_h;


import java.lang.foreign.MemorySegment;
import java.lang.ref.Cleaner;


public class OwnedStruct {
    var byte count;
    

    MemorySegment internal;

    private OwnedStruct() {
    }

    static OwnedStruct fromSegment(MemorySegment segment) {
        returnVal.internal = segment;
        var returnVal = new MyStruct();
        returnVal.count = dev.diplomattest.somelib.ntv.count(segment);
        
        return returnVal;
    }
    
}


============================
package dev.diplomattest.somelib;

import dev.diplomattest.somelib.ntv.somelib_h;


import java.lang.foreign.MemorySegment;
import java.lang.ref.Cleaner;


public class OwningStruct {
    var OwnedStruct a;
    var OwnedStruct b;
    

    MemorySegment internal;

    private OwningStruct() {
    }

    static OwningStruct fromSegment(MemorySegment segment) {
        returnVal.internal = segment;
        var returnVal = new MyStruct();
        returnVal.a = dev.diplomattest.somelib.ntv.a(segment);
        returnVal.b = dev.diplomattest.somelib.ntv.b(segment);
        
        return returnVal;
    }
    
    public static Opaque doOpaqueStuff(OwnedStruct input) {
        var inputNative = input.internal;
        var nativeVal = somelib_h.OwningStruct_do_opaque_stuff(inputNative);
        var returnVal = new Opaque();
        returnVal.internal = nativeVal;
        var cleaner = new Opaque.OpaqueCleaner(nativeVal);
        returnVal.cleanable = Main.cleaner.register(returnVal, cleaner);
        return returnVal;
    }
    
}


============================
package dev.diplomattest.somelib;

import dev.diplomattest.somelib.ntv.*;


import java.lang.foreign.MemorySegment;
import java.lang.ref.Cleaner;


public class Float64Vec {

    MemorySegment internal;
    Cleaner.Cleanable cleanable;

    static class Float64VecCleaner implements Runnable {

        MemorySegment segment;
        Float64VecCleaner(MemorySegment segment) {
            this.segment = segment;
        }

        public void run() {
            somelib_h.Float64Vec_destroy(this.segment);
        }
    }

    Float64Vec() {}
    
    public static Float64Vec new_(double[] v) {
        try (var arena = Arena.ofConfined()) {
            
            var nativeVal = somelib_h.Float64Vec_new();
            var returnVal = new Float64Vec();
            returnVal.internal = nativeVal;
            var cleaner = new Float64Vec.Float64VecCleaner(nativeVal);
            returnVal.cleanable = Main.cleaner.register(returnVal, cleaner);
            return returnVal;
        }
    }
    
    
    public double[] asBoxedSlice() {
        
        var nativeVal = somelib_h.Float64Vec_as_boxed_slice(internal);
        
    }
    
    public double[] asSlice() {
        
        var nativeVal = somelib_h.Float64Vec_as_slice(internal);
        
    }
    
}
============================
package dev.diplomattest.somelib;

import dev.diplomattest.somelib.ntv.*;


import java.lang.foreign.MemorySegment;
import java.lang.ref.Cleaner;


public class Opaque {

    MemorySegment internal;
    Cleaner.Cleanable cleanable;

    static class OpaqueCleaner implements Runnable {

        MemorySegment segment;
        OpaqueCleaner(MemorySegment segment) {
            this.segment = segment;
        }

        public void run() {
            somelib_h.Opaque_destroy(this.segment);
        }
    }

    Opaque() {}
    
    public static Opaque new_() {
        var nativeInvoker = somelib_h.Opaque_new.makeInvoker();
        var nativeVal = nativeInvoker.apply();
        var returnVal = new Opaque();
        returnVal.internal = nativeVal;
        var cleaner = new Opaque.OpaqueCleaner(nativeVal);
        returnVal.cleanable = Main.cleaner.register(returnVal, cleaner);
        return returnVal;
    }
    
    public static Opaque fromStr(String input) {
        try (var arena = Arena.ofConfined()) {
            var inputMemSeg = arena.allocateFrom(input, StandardCharsets.UTF_8);
            var inputLen = inputMemSeg.byteSize();
            var nativeVal = somelib_h.Opaque_from_str(inputMemSeg, inputLen - 1);
            var returnVal = new Opaque();
            returnVal.internal = nativeVal;
            var cleaner = new Opaque.OpaqueCleaner(nativeVal);
            returnVal.cleanable = Main.cleaner.register(returnVal, cleaner);
            return returnVal;
        }
    }
    
    public static long returnsUsize() {
        var nativeInvoker = somelib_h.Opaque_returns_usize.makeInvoker();
        var nativeVal = nativeInvoker.apply();
        return nativeVal;
    }
    
    public static OwningStruct returnsImported() {
        var nativeInvoker = somelib_h.Opaque_returns_imported.makeInvoker();
        var nativeVal = nativeInvoker.apply();
        return new OwningStruct(nativeVal);
    }
    
    
    public long internalLen() {
        
        var nativeVal = somelib_h.Opaque_internal_len(internal);
        return nativeVal;
    }
    
    public String getDebugStr() {
        
        var writeable = somelib_h.diplomat_buffer_write_create(0);
        somelib_h.Opaque_get_debug_str(internal, writeable);
        var buffer = DiplomatWrite.buf(writeable);
        var string = buffer.getString(0, StandardCharsets.UTF_8);
        somelib_h.diplomat_buffer_write_destroy(writeable);
        return string;
    }
    
}
============================
